### PR TITLE
Add SphericalHarmonic basis

### PR DESCRIPTION
**Pull request by isha1810**
## Proposed changes

Adds SphericalHarmonic as a basis to be used for generating the connectivity for S2 topology (apparent horizon connectivity).

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/NumericalAlgorithms/Spectral/Mesh.hpp [hunk 1]
```diff
@@ -17,6 +17,8 @@
 #include "Utilities/TypeTraits.hpp"  // IWYU pragma: keep
 #include "Utilities/TypeTraits/IsInteger.hpp"
 
+#include "Utilities/ErrorHandling/Assert.hpp"
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/Mesh.hpp]: I think this can be grouped in with the includes above, no?

**Review by nilsdeppe — COMMENTED** [review]
LGTM, just one tiny thing. Please go ahead and rebase on the latest develop and squash things into one commit :)

**Review by nilsdeppe — APPROVED** [review]

File: src/NumericalAlgorithms/Spectral/Mesh.hpp [hunk 2]
```diff
@@ -16,6 +16,7 @@
 #include "Utilities/Requires.hpp"
 #include "Utilities/TypeTraits.hpp"  // IWYU pragma: keep
 #include "Utilities/TypeTraits/IsInteger.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/Mesh.hpp]: It looks like alphabet is wrong here

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
can you squash the commits into one?

**Review by nilsdeppe — APPROVED** [review]
@kidder can you please review this?

**Review by kidder — APPROVED** [review]

**Pull request merged**